pkg/dockerregistry/server/manifestservice.go

```diff
@@ -235,3 +229,47 @@ func (m *manifestService) Delete(ctx context.Context, dgst digest.Digest) error
 	}
 	return nil
 }
+
+// manifestInflight tracks currently downloading manifests
+var manifestInflight = make(map[digest.Digest]struct{})
+
+// manifestInflightSync protects manifestInflight
+var manifestInflightSync sync.Mutex
+
+func (m *manifestService) migrateManifest(ctx context.Context, image *imageapi.Image, dgst digest.Digest, manifest distribution.Manifest, isLocalStored bool) {
```

Comment:
Could the method stop early if `!isLocalStored && len(image.DockerImageManifest) == 0` so that no synchronization and thread spawning needs to be done?